datamodel/samples.d/generate_timestampdefinitions.py

```diff
@@ -0,0 +1,1188 @@
+        EST_REQ_PLN,  # ATS-Cargo Ops comes later (different port call part)
+        ['CRGO'],
+        [NULL_VALUE, 'ALGS'],
+        "Services Planning",
+        'jit1_0',
+        is_cancelable=True,
+        need_vessel_position_for=EST_PLN,
+        need_event_location_for=REQ_PLN_ACT,
+    )
+
+    # ETS-Towage (Inbound)
+    xty_service_timestamps(
+        EST_REQ_PLN,  # ATS-Cargo Ops comes later (different port call part)
+        ['TOWG'],
+        [NULL_VALUE, 'INBD'],
+        "Services Planning",
+        'jit1_0',
+        include_phase_in_name=True,
+        is_cancelable=True,
+        need_vessel_position_for=EST_PLN,
+        need_event_location_for=REQ_PLN_ACT,
+    )
+
+    # ETS-Pilotage (Inbound)
+    xty_service_timestamps(
+        EST_REQ_PLN,  # ATS-Cargo Ops comes later (different port call part)
+        ['PILO'],
+        [NULL_VALUE, 'INBD'],
+        "Services Planning",
+        'jit1_0',
+        include_phase_in_name=True,
+        is_cancelable=True,
+        need_vessel_position_for=EST_PLN,
+        need_event_location_for=REQ_PLN_ACT,
+    )
+
+    # ETS-Mooring (Inbound) UC 19 - 24
+    xty_service_timestamps(
+        EST_REQ_PLN,  # ATS-Cargo Ops comes later (different port call part)
+        ['MOOR'],
+        [NULL_VALUE, 'INBD'],
+        "Services Planning",
+        'jit1_2',
+        include_phase_in_name=True,
+        is_cancelable=True,
+        need_vessel_position_for=EST_PLN,
+        need_event_location_for=REQ_PLN_ACT,
+    )
+
+    # Bunkering UC 25 - 30
+    xty_service_timestamps(
+        EST_REQ_PLN,
+        ['BUNK'],
+        [NULL_VALUE, 'ALGS'],
+        "Services Planning",
+        'jit1_2',
+        include_phase_in_name=False,
+        is_cancelable=True,
+        need_event_location_for=ALL_EVENT_CLASSIFIER_CODES,
+    )
+
+    # ETA PBP UC 31-33 + 35
+    generic_xty_timestamps(
+        PUBLISHER_PATTERN_CA2ATH,
+        ALL_EVENT_CLASSIFIER_CODES,
+        ['ARRI'],
+        ['PBPL'],
+        [NULL_VALUE, 'INBD'],
+        "Pilot Boarding Place Arrival Planning And Execution, Berth Arrival Execution",
+        'jit1_0',
+        need_event_location_for=ALL_EVENT_CLASSIFIER_CODES,
+    )
+
+    # EOSP UC 34
+    generate_special_timestamp(
+        'ESOP',
+        PUBLISHER_PATTERN_CA2ATH,
+        'ARRI',
+        'BRTH',
+        'INBD',
+        'Pilot Boarding Place Arrival Planning And Execution, Berth Arrival Execution',
+        'jit1_1',
+        need_vessel_position=True,  # FIXME: IFS says Optional, we do not support "required" or "absent".
+    )
+
+    # ATS-Pilotage (Inbound) UC 36 + 40
+    xty_service_timestamps(
+        ACT,
+        ['PILO'],
+        [NULL_VALUE, 'INBD'],
+        "Pilot Boarding Place Arrival Planning And Execution, Berth Arrival Execution",
+        'jit1_0',
+        include_phase_in_name=True,
+        is_cancelable=True,
+        need_vessel_position_for=ACT,
+        need_event_location_for=ACT,
+    )
```

Comment:
Might need to confirm that one with Charles but `ACT` means it `has happened`, so I suspect you cannot cancel it at that point (rinse and repeat for all `ACT`). In hindsight, I should have mentioned this as a part of my rule of thumb for timestamps.

Reply:
adjusted all service timestamp with eventclassifiercode ACT to is_cancelable=False